File: pie_lite/scoring.py
import re
from .config import PLACEHOLDER_RATES, COMPLEXITY_BONUS, FLOOR_AREA_DEFAULTS


def infer_build_type(description: str) -> str:
    """Simple keyword matching to guess build type."""
    text = description.lower()
    if any(phrase in text for phrase in ("new build", "new dwelling", "new housing")):
        return "new_build"
    if "loft" in text and "conversion" in text:
        return "loft_conversion"
    if "extension" in text:
        return "extension"
    if any(w in text for w in ("refurbishment", "renovation", "refurb")):
        return "refurbishment"
    return "other"


def extract_floor_area(description: str, build_type: str) -> tuple[float, str, str]:
    """
    Try to find a number followed by sqm/m².
    Returns (area, source, confidence).
    """
    match = re.search(
        r"(\d+(?:\.\d+)?)\s?(?:sq\.?\s?m|square\s?metres|m\s?²|sqm)",
        description,
        re.IGNORECASE,
    )
    if match:
        area = float(match.group(1))
        return area, "keyword_extraction", "low"
    default = FLOOR_AREA_DEFAULTS.get(build_type, 80)
    return float(default), "default", "low"


def calculate_opportunity_score(build_type: str, floor_area: float) -> int:
    """Transparent scoring formula."""
    bonus = COMPLEXITY_BONUS.get(build_type, 0)
    score = floor_area / 2.0 + bonus
    return min(100, int(round(score)))


def calculate_estimated_value(build_type: str, floor_area: float) -> float:
    """Uses placeholder rate – requires professional validation."""
    rate = PLACEHOLDER_RATES.get(build_type, 2000)
    return floor_area * rate

File: server/pie/scoring.ts
export type BuildType = 'extension' | 'new_build' | 'loft_conversion' | 'refurbishment' | 'other';

const PLACEHOLDER_RATES: Record<BuildType, number> = {
  extension: 2500,
  new_build: 3200,
  loft_conversion: 2200,
  refurbishment: 1800,
  other: 2000,
};

const COMPLEXITY_BONUS: Record<BuildType, number> = {
  extension: 10,
  new_build: 30,
  loft_conversion: 15,
  refurbishment: 5,
  other: 0,
};

const FLOOR_AREA_DEFAULTS: Record<BuildType, number> = {
  extension: 25,
  new_build: 120,
  loft_conversion: 40,
  refurbishment: 80,
  other: 80,
};

export function inferBuildType(description: string): BuildType {
  const text = description.toLowerCase();
  if (text.includes('new build') || text.includes('new dwelling') || text.includes('new housing')) {
    return 'new_build';
  }
  if (text.includes('loft') && text.includes('conversion')) {
    return 'loft_conversion';
  }
  if (text.includes('extension')) {
    return 'extension';
  }
  if (text.includes('refurbishment') || text.includes('renovation') || text.includes('refurb')) {
    return 'refurbishment';
  }
  return 'other';
}

export function extractFloorArea(
  description: string,
  buildType: BuildType,
): { area: number; source: 'keyword_extraction' | 'default'; confidence: 'low' } {
  const match = description.match(
    /(\d+(?:\.\d+)?)\s?(?:sq\.?\s?m|square\s?metres|m\s?²|sqm)/i,
  );
  if (match) {
    return { area: parseFloat(match[1]), source: 'keyword_extraction', confidence: 'low' };
  }
  return { area: FLOOR_AREA_DEFAULTS[buildType], source: 'default', confidence: 'low' };
}

export function calculateOpportunityScore(buildType: BuildType, floorArea: number): number {
  const bonus = COMPLEXITY_BONUS[buildType];
  const score = Math.floor(floorArea / 2 + bonus);
  return Math.min(100, score);
}

export function calculateEstimatedValue(buildType: BuildType, floorArea: number): number {
  const rate = PLACEHOLDER_RATES[buildType];
  return floorArea * rate;
}
